### ISSUE TITLE
Difference Between Azure.Cosmos 4.0.0 preview and Microsoft.Azure.Cosmos v3 SDK

### ISSUE THREAD

**Issue by vithal-wadje** [Azure/azure-cosmos-dotnet-v3#2316]
I seen there are two SDK's  for cosmos DB which are mentioned in the title but I observed both the SDK's source code redirected to the same GitHub repository and both SDK has same methods to perform the operations on cosmos DB, Then my question is what is the difference between these two SDK's . there is no proper documentation on same. same copy paste v3 example for v4 just difference in document is that its shown which version to be install.

Please explain on the same & update the respective document accordingly

**Comment by j82w** [comment]
V3 SDK is the current code in the master branch on github. This is the current primary SDK that all development and feature work is being done. This is the recommended SDK for all development.

V4 SDK is in the [v4 branch](https://github.com/Azure/azure-cosmos-dotnet-v3/tree/v4). The v4 SDK public surface area is based off the [Azure Core teams guidelines](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/Azure.Core). The interfaces are very similar to the v3 SDK but they are different. The v4 preview SDK was done to test out the new interfaces. Currently the v4 SDK is work is blocked on service related features. It's recommended to use the v3 SDK currently. Once we have a better estimate on time frame for the v4 SDK we will post an issue with more information.

**Comment by vithal-wadje** [comment]
Thanks for clarifying the my doubt . Hope new SDK will not be have just name changes of classes and methods , it should have some value additions such as Bulk insert etc.


Again thanks for your time

**Comment by ealsur** [comment]
@vithal-wadje V4 is currently in experimental phase, but ideally it would have the same features as V3, plus new ones (like AsyncEnumerable support). The [README of V4](https://github.com/Azure/azure-cosmos-dotnet-v3/tree/v4#query-items) has some feature examples.

**Comment by danielmarbach** [comment]
@ealsur @hansmbakker I was trying to find traces back to the v4 version changes. It seems the development has stalled since preview 3 and the preview was removed from the docs website. Has the move to Azure.Cosmos under the SDK unification attempts been stopped, halted, or is it in any other state? Would be great to share some insights.

**Comment by danielmarbach** [comment]
FYI this is my research so far

- It looks like the development on the [v4 branch](https://github.com/Azure/azure-cosmos-dotnet-v3/tree/v4) has stalled since November 2021
- [There was no update after the preview 3 on Nuget](https://www.nuget.org/packages/Azure.Cosmos/4.0.0-preview3)
- [Quickstart: Azure Cosmos DB SQL API client library for .NET](https://learn.microsoft.com/en-us/azure/cosmos-db/sql/quickstart-dotnet) no longer mentions the preview. The [initial commit](https://github.com/MicrosoftDocs/azure-docs/commit/8b1e63e0c56122649ecee347c2e4a53eb8b5dfea#diff-516dd001236cc4ce404f6505113d79537c39744e5b12f04f59b375eae563b395R100) is entirely built around the ` Microsoft.Azure.Cosmos` package and has [deleted the previous v4 preview page](https://github.com/MicrosoftDocs/azure-docs/commit/8b1e63e0c56122649ecee347c2e4a53eb8b5dfea#diff-601cc37bcae33d061897cd7c4ca2aa7adb307a673b39f254f724611f15a1a3b6) (see [original issue](https://github.com/MicrosoftDocs/azure-docs/issues/73357) that mentioned the sample should be marked as preview)
- The [v4 issue in the Azure SDK repo](https://github.com/Azure/azure-sdk-for-net/issues/12514#issuecomment-867415017) has remained unanswered
- [Difference Between Azure.Cosmos 4.0.0 preview and Microsoft.Azure.Cosmos v3 SDK](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/2316) has been closed and not updated since

Based on the public traces I could find, my observations would lead towards thinking the efforts have been abandoned.

**Comment by bartelink** [comment]
>has stalled since November 2021

Actually it was July/August 2020 - the last commit was only to put up the note.

TL;DR that V4 branch should be treated as a spike whose completion is not on the roadmap.

Do you have a particular aim in mind beyond not building new stuff on stuff that's about to be deprecated?

(I'm just a watcher - I sunk some time into targeting V4, and am sold on the general Azure.Core API design principles but ultimately the V3 nugets being stable is a perfectly fine state of affairs - ultimately the fragmentation that will happen whenever any V4 transition is actually triggered definitely has to provide benefits to outweigh the pain)

**Comment by FabianMeiswinkel** [comment]
@bartelink is absolutely correct - we have stopped the work on the V4 SDK version because the goal to comply to the Azure.Core design principles would require breaking changes - and convincing customers to move to new SDK versions with breaking changes is a hard sell - especially when the previous major version is "good enough". The maintenance cost for any parallel major version is too high to justify this. So, we are striving to have a new major version of the SDK roughly when the V2 SDK which has been deprecated in the meantime but still needs to stay supported for about two years gets out-of-support. With the new major version we are also stiving to move some of the non-trivial business logic around routing (sharding framework identifying shard and identifying and routing to the right replica) and retry policies in the direct transport stack into the service layer. We are evaluating different options there - because of course we must maintain the latency SLA boundaries only provided when using the direct transport today even when trimming down the SDK and moving logic into the service. But high-level that is the main goal and idea: Azure.Core alignment and slimming down the SDK by moving logic into a service endpoint without perf regressions.

**Comment by danielmarbach** [comment]
Thanks @bartelink and @FabianMeiswinkel for chiming in.

> Do you have a particular aim in mind beyond not building new stuff on stuff that's about to be deprecated?

No, the question was more coming from the angle that we want to track changes in the ecosystem and make sure we follow them, do early experiments so that we can adjust our offerings to the new clients within reasonable time frames ideally before all our customers are asking us "when do you support new client X" ;)

> and am sold on the general Azure.Core API design principles but ultimately the V3 nugets being stable is a perfectly fine state of affairs - ultimately the fragmentation that will happen whenever any V4 transition is actually triggered definitely has to provide benefits to outweigh the pain

:100: aligned.

> With the new major version we are also stiving to move some of the non-trivial business logic around routing (sharding framework identifying shard and identifying and routing to the right replica) and retry policies in the direct transport stack into the service layer.

That is a great plan because it also eases the requirements for all the other languages that need to be supported by the SDK libs.  Thanks for the transparency!
